Configurations/50-cppbuilder.conf

```diff
@@ -0,0 +1,37 @@
+my %targets = (
+    "BC-32" => {
+        inherit_from     => [ "BASE_Windows" ],
+        sys_id           => "WIN32",
+        bn_ops           => "BN_LLONG",
+        dso_scheme       => "win32",
+        thread_scheme    => "winthreads",
+        cc               => "bcc32c",
+        defines          => add("WIN32_LEAN_AND_MEAN", "OPENSSL_SYS_WIN32",
+                                "L_ENDIAN", "DSO_WIN32", "_stricmp=stricmp",
+                                "_strnicmp=strnicmp", "_malloca=malloc",
+                                "_freea=free", "_setmode=setmode"),
+        cflags           => picker(default => "-q -c -tWC -tWM",
+                                   debug   => "-Od -v -vi- -D_DEBUG",
+                                   release => "-O2"),
+        coutflag         => "-o",
+        LD               => "ilink32",
+        bin_lflags       => combine(picker(default => "-ap -Tpe -x -Gn -q",
+                                           debug   => '-j"$(BDS)\lib\win32c\debug" ' .
+                                                      '-L"$(BDS)\lib\win32c\debug"',
+                                           release => '-j"$(BDS)\lib\win32c\release" ' .
+                                                      '-L"$(BDS)\lib\win32c\release"'),
+                                    "wildargs.obj"),
+        ldoutflag        => ",",
+        ldpostoutflag    => ",,",
+        ld_resp_delim    => " +\n",
+        ex_libs          => add(sub {
+            my @ex_libs = ();
+            push @ex_libs, ("cw32mt.lib", "import32.lib", "crypt32.lib",
+                            "ws2_32.lib") unless $disabled{sock};
```

Comment:
This doesn't look right.  If OpenSSL gets configured with `no-sock`, it means that *none* of the libraries get added.  However, it's only "ws2_32.lib" that has to do with sockets.  Could it be that this push should really be these two?

``` perl
            push @ex_libs, ("cw32mt.lib", "import32.lib", "crypt32.lib");
            push @ex_libs, "ws2_32.lib" unless $disabled{sock};
```

Reply:
I took this from [what it was before](https://github.com/openssl/openssl/commit/8e56a4227bcd2e3eeb1cb18f3d0d5876ed76a8ce#diff-37c0fca57554a49849e973c9f5eed556a0dcab62e7ffbe12a0c10d448fa9ad8cL40-L42) (you'll need to click "Load diff" in `util/pl/BC-32.pl`). But you may be right, I will test the configuration options.

Three of these libraries are [documented with C++Builder](http://docwiki.embarcadero.com/RADStudio/Sydney/en/C_Run-Time_Library_Reference).

Reply:
Those lines in the old BC-32.pl look equally wrong, but ok, I can see how that was confusing.

Reply:
Turns out `crypt32.lib` is unnecessary in `no-sock` configuration, but `cw32mt.lib` and `import32.lib` are still needed.

Reply:
That's ... interesting.  I wonder if that's the same for the VC targets (which do link with `crypto32` unconditionally)

Reply:
Tried moving `crypt32.lib` to the `unless $disabled{sock}`-case in `VC-noCE-common`, but the dynamic build (`perl Configure VC-WIN32 --prefix=%CD% no-asm no-sock && nmake`) failed with:

```
   Creating library libcrypto.lib and object libcrypto.exp
libcrypto-shlib-bio_lib.obj : error LNK2019: unresolved external symbol _BIO_do_connect referenced in function _BIO_do_connect_retry
libcrypto-3.dll : fatal error LNK1120: 1 unresolved externals
```

while the static build (`perl Configure VC-WIN32 no-asm --prefix=%CD% no-asm no-shared no-sock && nmake`) failed with:

```
   Creating library providers\legacy.lib and object providers\legacy.exp
libcrypto.lib(libcrypto-lib-bio_lib.obj) : error LNK2019: unresolved external symbol _BIO_do_connect referenced in function _BIO_do_connect_retry
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2019: unresolved external symbol __imp__CertOpenStore@20 referenced in function _capi_open_store
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2019: unresolved external symbol __imp__CertCloseStore@8 referenced in function _capi_find_key
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2019: unresolved external symbol __imp__CertEnumCertificatesInStore@8 referenced in function _capi_find_cert
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2019: unresolved external symbol __imp__CertFindCertificateInStore@24 referenced in function _capi_find_cert
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2019: unresolved external symbol __imp__CertDuplicateCertificateContext@4 referenced in function _capi_load_ssl_client_cert
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2019: unresolved external symbol __imp__CertFreeCertificateContext@4 referenced in function _capi_dsa_free
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2019: unresolved external symbol __imp__CertGetCertificateContextProperty@16 referenced in function _capi_cert_get_fname
providers\legacy.dll : fatal error LNK1120: 8 unresolved externals
```

Tried them on the master branch (commit 5658470ce7) on separate Git worktrees, with `git submodule update --init --recursive && git clean -fxd` beforehand (if that matters). I used:

 - The "x86 Native Tools Command Prompt for VS 2019" window of Visual Studio Community 2019 version 16.8.1.
 - Git for Windows version 2.28.0.windows.1.

So for now I'm inclined to leave the VC targets as-is.

Reply:
I suspected we'd find something, and yeah, the CAPI engine does use crypt32 functions.  It's surprising that the BC config target differs...  could it be that engines aren't built?

Reply:
The CAPI engine is built (once I finished the DLL-build configuration), but it turns out that the `import32.lib`  provided with C++Builder contains the CryptoAPI symbols. This can be checked with the [`tdump.exe` utility](http://docwiki.embarcadero.com/RADStudio/Sydney/en/TDUMP.EXE,_the_File_Dumping_Utility):

```
>tdump %BDS%\lib\win32c\release\import32.lib | grep -B3 -A 1 CertOpenStore
171E32 COMENT  Purge: Yes, List: Yes, Class: 160 (0A0h), SubClass: 1 (01h)
    Dynamic link import (IMPDEF)
      Imported by: name
      Internal Name: CertOpenStore
      Module Name: CRYPT32.dll
```